text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+platform, since existing code would get a heads-up that it may not longer be
+compatible with all mainstream platforms.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are several potential drawbacks to the approach of this RFC:
+
+* It adds a significant level of pedanticness about portability to Rust.
+* It does not provide airtight guarantees.
+* It may create compiler performance issues, due to the use of SAT solving.
+
+The fact that it's a lint offers some help with the first two points; the use of
+`std` as a default portability level should also help quite a bit with
+pedanticness.
+
+The worry about SAT solving is harder to mitigate; there's not much concrete
+evidence in either direction. But it is yet another place where the fact that
+it's a lint could help: we may be able to simply skip checking pathological
+cases, if they indeed arise in practice. In any case, it's hard to know how
+concerned to be until we try it.
+
+While the fact that it's a lint gives us more leeway to experiment, it's also a
+lint that could produce widespread warnings throughout the ecosystem, so we need
+to exercise care.
+
+# Alternatives
+[alternatives]: #alternatives
+
+The main alternatives are:
+
+- **Give up on encouraging "portability by default"**, and instead just land
+  APIs in their natural location using today's `cfg` system. This is certainly
+  the less costly way to go. It's also *forward-compatible* with implementing
+  the proposed lint, so we should discuss the possibility of landing APIs under
+  `cfg` even before the lint is implemented.
+
+- **Use a less precise checking strategy.** In particular, rather than trying to
+  compare portabilities in a detailed, item-level way, we might just require
+  some crate-level "opt in". That could either take the form of acknowledging
+  "this code makes assumptions beyond the mainstream platform", or might list
+  the specific `cfg` assumptions the code is allowed to make. Of course, the
+  downside is that you get much less help making sure that your APIs are
+  properly labeled in place.
+
+# How we teach this
+[how-we-teach-this]: #how-we-teach-this
+
+For people simply using libraries, this feature "teaches itself" by generating
+warnings. Those warnings should make clear what to do to fix the problem, and
+ideally provide extended error information that describes the system in more
+detail.
+
+For library authors, the documentation for `cfg` and `match_cfg` would explain
+the implications for the lint, and walk through several examples illustrating
+the scenarios that arise in practice.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### Trait system
+
+Should we aim for better integration with the trait system? While we could
+probably catch relatively simple cases (where a trait is being used in a
+monomorphic way), generics raise tricky questions:
+
+```rust
+trait Foo {
+    fn foo();
+}
+
+struct MyType;
+
+impl Foo for MyType {
+    #[cfg(unix)]
```

Comment:
Oh, I was unclear I meant the `where cfg(x)` as "was kept because `cfg(x)` holds", not doing the actual gating on the fly, but remembering it was done. Though I should probably read the whole RFC before commenting, shouldn't I? 😆